Electron Desktop App

davidepatti opened the pull request:
This PR aims to track the finalization of a simple script to create macos/win/linux Electron desktop apps of the WebUI edumips64. 

Current approach:

- This should NOT be intended as an alternative/different version of the WebUI Edumips64 to be maintained: they are the SAME exact files produced when building the WebUI, just packaged to run without node.js or connecting to the Internet.
- The result of this PR should be something that does not interfere with the usual development and building processes. You want the Electron app? run the script. You don't care about it ? You will never notice it.

In summary, maybe having an offline version of WebUI represents a corner case (e.g. exams ?), but this PR can still provide feedback and updates on the current status for anyone interested in it.

**HOW IT WORKS**

1) Do a normal "npm build" 
2) Just launch "gen_webapp_macos.sh"
3) All the stuff produced in build/edumips64/build/gwt/war/edumips64 is packaged along with an Express server that serves the files, and an Electron App window the replaces the browser. They communicate locally on localhost port 8081 (configurable port), to not interfere with normal  "npm start", webpack, node.js testing stuff.
4) go to the folder dist/WebEduMips64-darwin-arm64  and click the app.

See electron_readme.txt for details and TODOs.

davidepatti commented:
Here's what it looks like... (time for new icon set?)
![Screenshot 2025-04-02 at 17 36 26](https://github.com/user-attachments/assets/0deea12e-5d78-4f9c-8dda-0defbac4d2f5)

davidepatti commented on electron_main.js:
```diff
@@ -0,0 +1,32 @@
+app.on('ready', createWindow);
+
+app.on('window-all-closed', () => {
+  if (process.platform !== 'darwin') {
```
It's kind an OS-related semantic of having all windows closed. In macos when you click "X" and close all windows the bar app icon remain active. Anyway, I'll check this stuff when I create the two other versions for Linux and Windows... we'll see...

lupino3 reviewed (changes requested):
The changes look good to me, I think we should test it a bit and then integrate it in the CI pipeline so we test it regularly, but in a subsequent PR. The approach here is good as-is.

The main change I'd like to ask is to move files away from the root directory. You could move everything to an `electron` directory for now, if that's ok.

davidepatti commented:
Moved files to electron, testing on windows.

davidepatti commented:
I've just found that powershell scripts should have .ps1 extension, not .sh, otherwise they are executed as bash in windows linux subsystem. So, I renamed the poweshell script, the app is created but when I run a get some errors. See the attached image, my only windows testing environment is a VM on mac running arm64 windows (I changed the target, of course).
<img width="916" alt="Screenshot 2025-04-03 at 19 26 31" src="https://github.com/user-attachments/assets/4c6a1d44-6d25-4dc2-ac18-abe971b72849" />

davidepatti commented:
Confirmed working scripts for win32,linux,macos!
